File: src/knowledge.ts
import { createHash } from "node:crypto";

import type { EmbeddingProvider } from "./retrieval.js";
import type { ResearchStore } from "./store.js";
import { normalizeArabic } from "./text.js";

const lexicon = [
  ["المبادرة المصرية للحقوق الشخصية", "organization", ["المبادرة المصرية"]],
  ["النيابة العامة", "organization", []],
  ["مجلس النواب", "organization", ["البرلمان المصري"]],
  ["مجلس الوزراء", "organization", ["رئاسة مجلس الوزراء"]],
  ["وزارة الداخلية", "organization", []],
  ["وزارة التموين", "organization", []],
  ["وزارة التضامن الاجتماعي", "organization", []],
  ["وزارة الخارجية", "organization", []],
  ["وزارة التعليم العالي", "organization", []],
  ["محكمة النقض", "organization", []],
  ["المحكمة الدستورية العليا", "organization", []],
  ["المجلس القومي لحقوق الإنسان", "organization", []],
  ["مفوضية الأمم المتحدة لشؤون اللاجئين", "organization", ["المفوضية السامية للأمم المتحدة لشؤون اللاجئين"]],
  ["جامعة الدول العربية", "organization", []],
  ["جهاز مستقبل مصر", "organization", []],
  ["الجهاز المركزي للتعبئة العامة والإحصاء", "organization", ["التعبئة والإحصاء"]],
  ["القاهرة", "location", []], ["الإسكندرية", "location", []], ["سيناء", "location", []],
  ["رفح", "location", []], ["العريش", "location", []]
] as const;
const reportingVerbs = ["اعلن", "اعلنت", "تعلن", "قال", "تقول", "اكد", "يوكد", "تؤكد", "ذكر", "افاد", "صرح", "اوضحت", "اوضح", "كشف", "اشار", "نفى", "نفت", "تنفي", "يطالب", "طالبت", "حذر", "تحذر"];
const entityPrefixes: Array<[string, string]> = [
  ["وزاره", "organization"], ["مجلس", "organization"], ["محكمه", "organization"], ["جهاز", "organization"],
  ["هيئه", "organization"], ["مؤسسه", "organization"], ["منظمه", "organization"], ["جامعه", "organization"],
  ["مفوضيه", "organization"], ["نيابه", "organization"], ["لجنه", "organization"], ["رئاسه", "organization"]
];
const months: Record<string, number> = { يناير: 1, فبراير: 2, مارس: 3, ابريل: 4, مايو: 5, يونيو: 6, يوليو: 7, اغسطس: 8, سبتمبر: 9, اكتوبر: 10, نوفمبر: 11, ديسمبر: 12 };

export interface KnowledgeReport { documentId: number; entities: number; claims: number; events: number; embedded: boolean }

export class KnowledgeIndexer {
  constructor(private readonly store: ResearchStore, private readonly embeddingProvider?: EmbeddingProvider) {}

  indexDocument(documentId: number): KnowledgeReport {
    const document = this.store.getDocument(documentId);
    if (!document) throw new Error(`Unknown document: ${documentId}`);
    if (document.documentType === "excluded") return { documentId, entities: 0, claims: 0, events: 0, embedded: false };
    this.store.resetDocumentKnowledge(documentId);
    const text = `${document.title}\n${document.excerpt}\n${document.content ?? ""}`;
    const normalized = normalizeArabic(text);
    const entities: string[] = [];
    for (const [canonicalName, entityType, aliases] of lexicon) {
      const candidates = [canonicalName, ...aliases];
      const mentions = candidates.reduce((count, name) => count + occurrences(normalized, normalizeArabic(name)), 0);
      if (!mentions) continue;
      this.store.linkEntity(documentId, canonicalName, entityType, mentions, normalized.includes(normalizeArabic(canonicalName)) ? 0.95 : 0.8, [...aliases]);
      entities.push(canonicalName);
    }
    const known = new Set(lexicon.flatMap(([name]) => {
      const normalizedName = normalizeArabic(name);
      return [normalizedName, normalizedName.replace(/^ال/u, "")];
    }));
    for (const [canonicalName, entityType] of discoverEntities(text)) {
      if (known.has(normalizeArabic(canonicalName))) continue;
      const mentions = occurrences(normalized, normalizeArabic(canonicalName));
      if (mentions < 1) continue;
      this.store.linkEntity(documentId, canonicalName, entityType, mentions, 0.7);
      entities.push(canonicalName);
    }
    const claims = [...new Set((document.content ?? "").split(/[.!؟؛\n]+/).map((sentence) => sentence.replace(/\s+/g, " ").trim()).filter((sentence) => sentence.length >= 25 && reportingVerbs.some((verb) => normalizeArabic(sentence).includes(verb))))];
    for (const claim of claims) {
      const classification = classifyClaim(claim);
      this.store.upsertClaim(documentId, claim, classification.claimType, classification.stance, classification.confidence);
    }
    const eventAt = extractEventDate(text);
    if (eventAt) this.store.updateDocumentEventAt(documentId, eventAt);
    const locations = lexicon
      .filter(([name, type]) => type === "location" && normalized.includes(normalizeArabic(name)))
      .map(([name]) => name);
    this.store.upsertEventForDocument(documentId, {
      title: document.title, summary: document.excerpt || (document.content ?? "").slice(0, 500),
      occurredAt: eventAt ?? document.eventAt ?? document.publishedAt ?? document.archivedAt,
      eventType: document.topics[0] ?? document.documentType, location: locations[0] ?? null
    });
    this.store.purgeOrphanKnowledge();
    if (this.embeddingProvider) {
      this.store.upsertEmbedding(documentId, this.embeddingProvider.provider, this.embeddingProvider.model, this.embeddingProvider.embed(text), createHash("sha256").update(text).digest("hex"));
    }
    return { documentId, entities: entities.length, claims: claims.length, events: 1, embedded: Boolean(this.embeddingProvider) };
  }

  backfill(limit = 10_000): KnowledgeReport[] { return this.store.listDocumentIds(limit).map((id) => this.indexDocument(id)); }
}

function discoverEntities(value: string): Array<[string, string]> {
  const normalized = normalizeArabic(value);
  const found = new Map<string, string>();
  const stop = new Set(["تعلن", "اعلن", "اعلنت", "قال", "تقول", "بشأن", "بشان", "عن", "في", "من", "الى", "الي", "قرار", "قرارا", "بيان", "بيانا", "قررت", "يتناول", "تناول", "اخلاء", "سبيل", "بعدم", "الدعوى", "الدعوي", "والزمت", "لصالح", "على", "ان", "اليوم", "المقرر", "تنظر", "بحث", "يبحث", "شارك", "مناقشه", "مناقشة", "بشأن", "اجتماعا", "اجتماع", "رييس", "رئيس", "مركز", "المعلومات"]);
  for (const [prefix, entityType] of entityPrefixes) {
    const pattern = new RegExp(`${prefix}(?:\\s+[ء-ي]{2,}){1,4}`, "gu");
    for (const match of normalized.matchAll(pattern)) {
      const words = match[0]!.trim().split(" ").slice(0, 5);
      const stopAt = words.findIndex((word, index) => index > 0 && stop.has(word));
      const name = words.slice(0, stopAt > 0 ? stopAt : words.length).join(" ");
      if (name.split(" ").length >= 2 && name.length <= 90) found.set(name, entityType);
    }
  }
  return [...found.entries()];
}

function classifyClaim(value: string): { claimType: string; stance: string; confidence: number } {
  const normalized = normalizeArabic(value);
  if (/(نفى|نفت|تنفي|ينفي)/u.test(normalized)) return { claimType: "denial", stance: "contradicts", confidence: 0.82 };
  if (/(طالب|طالبت|يطالب)/u.test(normalized)) return { claimType: "demand", stance: "advocates", confidence: 0.8 };
  if (/(حذر|تحذر|يحذر)/u.test(normalized)) return { claimType: "warning", stance: "warns", confidence: 0.8 };
  return { claimType: "reported_statement", stance: "reports", confidence: 0.75 };
}

function extractEventDate(value: string): string | null {
  const normalized = normalizeArabic(value);
  const numeric = normalized.match(/\b(20\d{2})[-/](\d{1,2})[-/](\d{1,2})\b/u);
  if (numeric) return isoDate(Number(numeric[1]), Number(numeric[2]), Number(numeric[3]));
  const arabic = normalized.match(/\b(\d{1,2})\s+(يناير|فبراير|مارس|ابريل|مايو|يونيو|يوليو|اغسطس|سبتمبر|اكتوبر|نوفمبر|ديسمبر)\s+(20\d{2})\b/u);
  return arabic ? isoDate(Number(arabic[3]), months[arabic[2]!]!, Number(arabic[1])) : null;
}

function isoDate(year: number, month: number, day: number): string | null {
  const date = new Date(Date.UTC(year, month - 1, day));
  return date.getUTCFullYear() === year && date.getUTCMonth() === month - 1 && date.getUTCDate() === day ? date.toISOString() : null;
}

function occurrences(value: string, needle: string): number {
  if (!needle) return 0;
  let count = 0; let offset = 0;
  while ((offset = value.indexOf(needle, offset)) >= 0) { count += 1; offset += needle.length; }
  return count;
}

File: src/egypt_research_mcp/knowledge.py
from __future__ import annotations

import hashlib
import re

from .models import KnowledgeIndexReport
from .normalization import normalize_arabic
from .retrieval import LocalEmbeddingProvider
from .store import ResearchStore


_ENTITY_LEXICON = (
    (
        "المبادرة المصرية للحقوق الشخصية",
        "organization",
        ("المبادرة المصرية",),
    ),
    ("النيابة العامة", "organization", ()),
    ("مجلس النواب", "organization", ("البرلمان المصري",)),
    ("وزارة الداخلية", "organization", ()),
    ("محكمة النقض", "organization", ()),
    ("الجهاز المركزي للتعبئة العامة والإحصاء", "organization", ("التعبئة والإحصاء",)),
    ("القاهرة", "location", ()),
    ("الإسكندرية", "location", ()),
    ("سيناء", "location", ()),
    ("رفح", "location", ()),
    ("العريش", "location", ()),
)

_REPORTING_VERBS = (
    "اعلن",
    "قال",
    "اكد",
    "ذكر",
    "افاد",
    "صرح",
    "اوضحت",
    "اوضح",
)


class KnowledgeIndexer:
    def __init__(
        self,
        store: ResearchStore,
        *,
        embedding_provider: LocalEmbeddingProvider | None = None,
    ) -> None:
        self.store = store
        self.embedding_provider = embedding_provider

    def index_document(self, document_id: int) -> KnowledgeIndexReport:
        document = self.store.get_document(document_id)
        if document is None:
            raise ValueError(f"Unknown document: {document_id}")
        self.store.reset_document_knowledge(document_id)
        text = f"{document.title}\n{document.excerpt}\n{document.content}"
        normalized = normalize_arabic(text)

        entities = []
        for canonical_name, entity_type, aliases in _ENTITY_LEXICON:
            candidates = (canonical_name, *aliases)
            mentions = sum(normalized.count(normalize_arabic(name)) for name in candidates)
            if not mentions:
                continue
            self.store.link_entity(
                document_id,
                canonical_name,
                entity_type,
                mentions=mentions,
                confidence=0.95 if normalize_arabic(canonical_name) in normalized else 0.8,
                aliases=list(aliases),
            )
            entities.append(canonical_name)

        claims = []
        for sentence in re.split(r"[.!؟\n]+", document.content):
            sentence = " ".join(sentence.split()).strip()
            normalized_sentence = normalize_arabic(sentence)
            if len(sentence) < 25 or not any(
                verb in normalized_sentence for verb in _REPORTING_VERBS
            ):
                continue
            self.store.upsert_claim(document_id, sentence)
            claims.append(sentence)

        locations = [
            name
            for name, entity_type, _ in _ENTITY_LEXICON
            if entity_type == "location" and normalize_arabic(name) in normalized
        ]
        occurred_at = document.event_at or document.published_at or document.archived_at
        self.store.upsert_event_for_document(
            document_id,
            title=document.title,
            summary=document.excerpt or document.content[:500],
            occurred_at=occurred_at,
            event_type=document.topics[0] if document.topics else document.document_type,
            location=locations[0] if locations else None,
        )
        self.store.purge_orphan_knowledge()

        embedded = self.embedding_provider is not None
        if self.embedding_provider:
            vector = self.embedding_provider.embed(text)
            self.store.upsert_embedding(
                document_id,
                provider=self.embedding_provider.provider,
                model=self.embedding_provider.model,
                vector=vector,
                content_hash=hashlib.sha256(text.encode()).hexdigest(),
            )
        return KnowledgeIndexReport(
            document_id=document_id,
            entities=len(entities),
            claims=len(claims),
            events=1,
            embedded=embedded,
        )

    def backfill(self, *, limit: int = 10_000) -> list[KnowledgeIndexReport]:
        return [
            self.index_document(document_id)
            for document_id in self.store.list_document_ids(limit=limit)
        ]
